# Scikit-learn exploration

## Made by:
- Annisa Rahim / 13518089
- Stefanus Gusega Gunawan / 13518149

# Breast-Cancer Dataset

Load the datasets first.

In [1]:
import pandas as pd
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

load = load_breast_cancer()
cancer = pd.DataFrame(load.data, columns=load.feature_names)
target = pd.DataFrame(load.target, columns =['target'])

# Split 80 : 20
X_train, X_valid, y_train, y_valid = train_test_split(cancer, target, train_size = 0.8, test_size = 0.2, random_state=1)
X_train

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
408        17.99         20.66          117.80      991.7          0.10360   
4          20.29         14.34          135.10     1297.0          0.10030   
307         9.00         14.40           56.36      246.3          0.07005   
386        12.21         14.09           78.78      462.0          0.08108   
404        12.34         14.95           78.29      469.1          0.08682   
..           ...           ...             ...        ...              ...   
129        19.79         25.12          130.40     1192.0          0.10150   
144        10.75         14.97           68.26      355.3          0.07793   
72         17.20         24.52          114.20      929.4          0.10710   
235        14.03         21.25           89.79      603.4          0.09070   
37         13.03         18.42           82.61      523.8          0.08983   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
408           0.13040        0.120100             0.088240         0.1992   
4             0.13280        0.198000             0.104300         0.1809   
307           0.03116        0.003681             0.003472         0.1788   
386           0.07823        0.068390             0.025340         0.1646   
404           0.04571        0.021090             0.020540         0.1571   
..                ...             ...                  ...            ...   
129           0.15890        0.254500             0.114900         0.2202   
144           0.05139        0.022510             0.007875         0.1399   
72            0.18300        0.169200             0.079440         0.1927   
235           0.06945        0.014620             0.018960         0.1517   
37            0.03766        0.025620             0.029230         0.1467   

     mean fractal dimension  ...  worst radius  worst texture  \
408                 0.06069  ...        21.080          25.41   
4                   0.05883  ...        22.540          16.67   
307                 0.06833  ...         9.699          20.07   
386                 0.06154  ...        13.130          19.29   
404                 0.05708  ...        13.180          16.85   
..                      ...  ...           ...            ...   
129                 0.06113  ...        22.630          33.58   
144                 0.05688  ...        11.950          20.72   
72                  0.06487  ...        23.320          33.82   
235                 0.05835  ...        15.330          30.28   
37                  0.05863  ...        13.300          22.81   

     worst perimeter  worst area  worst smoothness  worst compactness  \
408           138.10      1349.0           0.14820            0.37350   
4             152.20      1575.0           0.13740            0.20500   
307            60.90       285.5           0.09861            0.05232   
386            87.65       529.9           0.10260            0.24310   
404            84.11       533.1           0.10480            0.06744   
..               ...         ...               ...                ...   
129           148.70      1589.0           0.12750            0.38610   
144            77.79       441.2           0.10760            0.12230   
72            151.60      1681.0           0.15850            0.73940   
235            98.27       715.5           0.12870            0.15130   
37             84.46       545.9           0.09701            0.04619   

     worst concavity  worst concave points  worst symmetry  \
408          0.33010               0.19740          0.3060   
4            0.40000               0.16250          0.2364   
307          0.01472               0.01389          0.2991   
386          0.30760               0.09140          0.2677   
404          0.04921               0.04793          0.2298   
..               ...                   ...             ...   
129          0.56730               0.17320         

Check whether there is missing values.

In [2]:
cols_missing = [col for col in cancer.columns if (cancer[col].isnull().any())]
cols_missing

[]

## Train datasets
### DecisionTreeClassifier

In [3]:
from sklearn.tree import DecisionTreeClassifier

classifier = DecisionTreeClassifier(random_state=1)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(random_state=1)

Print the tree.

In [4]:
from sklearn.tree import export_text

tree = export_text(classifier, feature_names = list(cancer.columns))
print(tree)

|--- worst perimeter <= 106.05
|   |--- worst concave points <= 0.16
|   |   |--- worst concave points <= 0.14
|   |   |   |--- area error <= 48.98
|   |   |   |   |--- class: 1
|   |   |   |--- area error >  48.98
|   |   |   |   |--- worst radius <= 13.55
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- worst radius >  13.55
|   |   |   |   |   |--- class: 1
|   |   |--- worst concave points >  0.14
|   |   |   |--- worst texture <= 29.45
|   |   |   |   |--- class: 1
|   |   |   |--- worst texture >  29.45
|   |   |   |   |--- class: 0
|   |--- worst concave points >  0.16
|   |   |--- worst texture <= 24.78
|   |   |   |--- class: 1
|   |   |--- worst texture >  24.78
|   |   |   |--- class: 0
|--- worst perimeter >  106.05
|   |--- worst texture <= 20.65
|   |   |--- worst perimeter <= 116.80
|   |   |   |--- class: 1
|   |   |--- worst perimeter >  116.80
|   |   |   |--- mean smoothness <= 0.08
|   |   |   |   |--- class: 1
|   |   |   |--- mean smoothness >  0.08
|   |   

Give prediction to training data so we can see how good the model is. Compared to true value of training data. 

In [5]:
predict_train = classifier.predict(X_train)

Evaluate the model with Accuracy Metrics.

$$
Accuracy = \frac {Number of Correct predictions}{Total number of predictions made}
$$

In [6]:
from sklearn.metrics import accuracy_score

acc_train_score = accuracy_score(y_train, predict_train)
acc_train_score

1.0

Evaluate the model with F1 score Metrics. The metrics used based on this confusion matrix. F1 Score is the **_Harmonic Mean_** between precision and recall.
<!-- ![Alt Text](image path "title") -->
![Alt Text](https://miro.medium.com/max/700/1*OhEnS-T54Cz0YSTl_c3Dwg.jpeg "Confusion Matrix")

$$
Precision = \frac {TruePositives}{TruePositives+FalsePositives}
$$ <br>
$$
Recall = \frac {TruePositives}{TruePositives+FalseNegatives}
$$ <br>
$$
F1 = 2*(\frac{1}{\frac {1}{Precision} + \frac {1}{Recall}}) = 2*(\frac {TruePositives}{2TruePositives+FalsePositives+FalseNegatives})
$$

In [7]:
from sklearn.metrics import f1_score

f1_train_score = f1_score(y_train, predict_train)
f1_train_score

1.0

Now predict the test data.

In [8]:
predict_valid = classifier.predict(X_valid)

Evaluate the model with Accuracy Metrics.

In [9]:
acc_valid_score = accuracy_score(y_valid, predict_valid)
acc_valid_score

0.9473684210526315

Evaluate the model with F1 Metrics

In [10]:
f1_valid_score = f1_score(y_valid, predict_valid)
f1_valid_score

0.9594594594594595

### Id3Estimator

Train the model and print the tree.

In [11]:
import six
import sys
sys.modules['sklearn.externals.six'] = six

from id3 import Id3Estimator, export

id3_model1 = Id3Estimator()

id3_model1.fit(X_train, y_train.values.ravel()) # make it numpy 1D array, not a df

# export text
tree_text1 = export.export_text(id3_model1.tree_, feature_names=cancer.columns)

print(tree_text1)



worst perimeter <=105.15
|   worst concave points <=0.14
|   |   radius error <=0.64: 1 (249) 
|   |   radius error >0.64
|   |   |   mean radius <=12.27: 0 (1) 
|   |   |   mean radius >12.27: 1 (2) 
|   worst concave points >0.14
|   |   worst texture <=25.94: 1 (6) 
|   |   worst texture >25.94
|   |   |   mean compactness <=0.12
|   |   |   |   mean radius <=14.06: 1 (2) 
|   |   |   |   mean radius >14.06: 0 (1) 
|   |   |   mean compactness >0.12: 0 (5) 
worst perimeter >105.15
|   worst concave points <=0.15
|   |   worst texture <=19.91: 1 (13) 
|   |   worst texture >19.91
|   |   |   worst radius <=16.80
|   |   |   |   mean smoothness <=0.09: 1 (8) 
|   |   |   |   mean smoothness >0.09
|   |   |   |   |   smoothness error <=0.00: 1 (2) 
|   |   |   |   |   smoothness error >0.00: 0 (5) 
|   |   |   worst radius >16.80
|   |   |   |   worst concavity <=0.21
|   |   |   |   |   mean texture <=21.26: 1 (2) 
|   |   |   |   |   mean texture >21.26: 0 (2) 
|   |   |   |   worst

Evaluate the model with Accuracy and F1 score metrics.

In [12]:
# Model Evaluation
predict_train1 = id3_model1.predict(X_train)

print("Accuracy: ", accuracy_score(predict_train1, y_train))
print("F1 score: ", f1_score(predict_train1, y_train))

Accuracy:  1.0
F1 score:  1.0


Evaluate the model to test data with Accuracy and F1 score metrics.

In [13]:
# Model Prediction
predict_test1 = id3_model1.predict(X_valid)

print("Accuracy: ", accuracy_score(predict_test1,y_valid))
print("F1 score: ", f1_score(predict_test1,y_valid))

Accuracy:  0.9385964912280702
F1 score:  0.953020134228188


### KMeans

Train the datasets.

In [14]:
from sklearn.cluster import KMeans

# Determine how many clusters
n_clust = y_train['target'].nunique()

kmeans_model_1 = KMeans(n_clusters=n_clust,random_state=1)
kmeans_model_1.fit(X_train)

KMeans(n_clusters=2, random_state=1)

Predict the data train and data test.

In [15]:
kmeans_predict_train = kmeans_model_1.predict(X_train)
kmeans_predict_test = kmeans_model_1.predict(X_valid)
print(kmeans_predict_train)
print(y_train)

[1 1 0 0 0 0 0 1 1 0 0 1 0 0 1 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0
 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0 1
 0 0 1 1 0 0 0 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 1 1 1 0 1 0 0 0 0 0 0 0 0
 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 1 0 0 1 1 0
 0 0 1 0 0 0 0 1 1 0 0 0 1 0 0 0 1 0 0 1 0 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 1
 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 1 0
 0 0 0 1 0 0 1 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 0 0 0 0 0 0 1
 1 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 1 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1
 1 1 0 0 0 1 1 0 0 0 0 1 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 1 1 0 0 0 0 0 0 1 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 1 0
 1 0 0 0 0 0 1 0 1 0 0]
     target
408       0
4         0
307       1
386       1
404       1
..  

Define a function to switch the value from 0 to 1

In [16]:
import numpy as np
def switch(x: int) -> int:
    if (x==0):
        return x+1
    elif (x==1):
        return x-1

For evaluation with training data.

In [17]:
# Find out each cluster belong with which class

# First clusterization using first prediction (k_means_train)
# Calculate the accuracy and F1
acc_train = accuracy_score(y_train, kmeans_predict_train)
f1_train = f1_score(y_train, kmeans_predict_train)
print(f"First clusterization accuracy: {acc_train}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_train = pd.Series(kmeans_predict_train)
fix = switched_train.apply(switch)
acc_train1= accuracy_score(y_train, fix)
f1_train2= f1_score(y_train, fix)
print(f"Second clusterization accuracy: {acc_train1}")
print(f"Second clusterization F1: {f1_train2}")

First clusterization accuracy: 0.15164835164835164
First clusterization F1: 0.005154639175257732
Second clusterization accuracy: 0.8483516483516483
Second clusterization F1: 0.891679748822606


The performance can be seen by the greatest metrics score. So the cluster 0 belongs to class of 1, cluster 1 belongs to class of 0. Now, for evaluation with validation data (datatest).

In [18]:
# First clusterization
acc_valid = accuracy_score(y_valid, kmeans_predict_test)
f1_valid = f1_score(y_valid, kmeans_predict_test)
print(f"First clusterization accuracy: {acc_valid}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_valid = pd.Series(kmeans_predict_test).apply(switch)
acc_valid_switch = accuracy_score(y_valid, switched_valid)
f1_valid_switch = f1_score(y_valid, switched_valid)
print(f"Second clusterization accuracy: {acc_valid_switch}")
print(f"Second clusterization F1: {f1_valid_switch}")

First clusterization accuracy: 0.19298245614035087
First clusterization F1: 0.005154639175257732
Second clusterization accuracy: 0.8070175438596491
Second clusterization F1: 0.8674698795180723


Same as when we evaluate with training data, the second clusterization is better.

### LogisticRegression

Scaling the dataset.

In [19]:
# Scaling, soalnya kl gapake scaling atau max iter ga ditambah 
# dia ada ConvergenceWarning
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_valid = scaler.fit_transform(X_valid)

Modelling the scaled dataset.

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score

model = LogisticRegression()

model.fit(X_train,y_train.values.ravel())

LogisticRegression()

Evaluate the model to train data with Accuracy and F1 score metrics.

In [21]:
predict_train = model.predict(X_train)

acc = accuracy_score(y_train, predict_train)
f1 = f1_score(y_train, predict_train)

print("Accuracy: ", acc)
print("F1 score: ", f1)

Accuracy:  0.9912087912087912
F1 score:  0.993006993006993


Evaluate the model to test data with Accuracy and F1 score metrics.

In [22]:
predict_test = model.predict(X_valid)

acc = accuracy_score(y_valid, predict_test)
f1 = f1_score(y_valid, predict_test)

print("Accuracy: ", acc)
print("F1 score: ", f1)

Accuracy:  0.9736842105263158
F1 score:  0.9793103448275863


no scaler & no max iter:
- Accuracy:  0.9473684210526315
- F1 score:  0.9589041095890412

max iter:
- Accuracy:  0.9473684210526315
- F1 score:  0.9594594594594595

with scaling 
- standard
    - Accuracy:  0.9736842105263158
    - F1 score:  0.9793103448275863
- minmax
    - Accuracy:  0.9736842105263158
    - F1 score:  0.9795918367346939

### Neutral_network

### SVM

Define the SVM (Support Vector Machines) constructor

In [71]:
from sklearn.svm import SVC

svm = SVC(random_state=1)

Train the traning dataset

In [74]:
svm.fit(X_train, y_train.values.ravel())
# .values will give the values in an array (shape: (n,1)
# .ravel will convert that array shape to (n, )

SVC(random_state=1)

Predict the dataset.

In [85]:
pred_train_svm_breast = svm.predict(X_train)
pred_val_svm_breast = svm.predict(X_valid)
pred_val_svm_breast

array([1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 1, 1])

Evaluate data train prediction with using Accuracy and F1 metrics

In [89]:
acc_score = accuracy_score(y_train, pred_train_svm_breast)
f1 = f1_score(y_train, pred_train_svm_breast)
print("Accuracy : ", acc_score)
print("F1 : ",f1)

Accuracy :  0.9868131868131869
F1 :  0.9895470383275261


Evaluate data test prediction with using Accuracy and F1 metrics

In [90]:
acc_score = accuracy_score(y_valid, pred_val_svm_breast)
f1 = f1_score(y_valid, pred_val_svm_breast)
print("Accuracy : ", acc_score)
print("F1 : ",f1)

Accuracy :  0.9649122807017544
F1 :  0.9722222222222222


# Play-Tennis Dataset

Load the dataset first.

In [24]:
tennis = pd.read_csv('datasets/PlayTennis.csv')
X_tennis = tennis.drop(['play'],axis=1)
y_tennis = pd.DataFrame(tennis['play'])

# Split 80 : 20
X_train_tennis, X_valid_tennis, y_train_tennis, y_valid_tennis = train_test_split(X_tennis, y_tennis, train_size = 0.8, test_size = 0.2, random_state=1)
y_valid_tennis

play
3  yes
7   no
6  yes

Check whether there are missing values

In [25]:
cols_with_missing_tennis = [col for col in tennis.columns if tennis[col].isnull().any()]
cols_with_missing_tennis

[]

## Train datasets
### DecisionTreeClassifier

Make the categorical variable to numeric variable

In [76]:
from sklearn.preprocessing import LabelEncoder

cond = (X_train_tennis.dtypes == 'object')
object_cols = list(cond[cond].index)

label_X_train = X_train_tennis.copy()
label_X_valid = X_valid_tennis.copy()

label_encoder = LabelEncoder()
for col in object_cols:
    label_X_train[col] = label_encoder.fit_transform(X_train_tennis[col])
    label_X_valid[col] = label_encoder.transform(X_valid_tennis[col])

# In case needed
label_y_train = label_encoder.fit_transform(y_train_tennis)
label_y_valid = label_encoder.transform(y_valid_tennis)

label_X_train

outlook  temp  humidity  windy
2         0     1         0  False
10        2     2         1   True
4         1     0         1  False
1         2     1         0   True
12        0     1         1  False
0         2     1         0  False
13        1     2         0   True
9         1     2         1  False
8         2     0         1  False
11        0     2         0   True
5         1     0         1   True

Train the model.

In [30]:
tennis_classifier = DecisionTreeClassifier(random_state=1)
tennis_classifier.fit(label_X_train, y_train_tennis)

DecisionTreeClassifier(random_state=1)

Print the tree.

In [31]:
tree_tennis = export_text(tennis_classifier, feature_names = list(X_tennis.columns))
print(tree_tennis)

|--- outlook <= 0.50
|   |--- class: yes
|--- outlook >  0.50
|   |--- humidity <= 0.50
|   |   |--- class: no
|   |--- humidity >  0.50
|   |   |--- windy <= 0.50
|   |   |   |--- class: yes
|   |   |--- windy >  0.50
|   |   |   |--- outlook <= 1.50
|   |   |   |   |--- class: no
|   |   |   |--- outlook >  1.50
|   |   |   |   |--- class: yes



Predict training data with the model.

In [32]:
predict_train_tennis = tennis_classifier.predict(label_X_train)
predict_train_tennis

array(['yes', 'yes', 'yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'yes',
       'no'], dtype=object)

Evaluate the model to training data with Accuracy Metrics

In [33]:
acc_train_tennis = accuracy_score(y_train_tennis,predict_train_tennis)
acc_train_tennis

1.0

Evaluate the model to training data with F1 Metrics

In [34]:
f1_train_tennis = f1_score(y_train_tennis,predict_train_tennis,pos_label='yes')
f1_train_tennis

1.0

Predict the test data using model.

In [35]:
predict_valid_tennis = tennis_classifier.predict(label_X_valid)
predict_valid_tennis

array(['no', 'no', 'yes'], dtype=object)

Evaluate the model with datatest using Accuracy Metrics

In [36]:
acc_valid_tennis = accuracy_score(y_valid_tennis,predict_valid_tennis)
acc_valid_tennis

0.6666666666666666

Evaluate the model with datatest using F1 Metrics

In [37]:
f1_valid_tennis = f1_score(y_valid_tennis,predict_valid_tennis,pos_label='yes')
f1_valid_tennis

0.6666666666666666

### Id3Estimator

In [38]:
id3_model2 = Id3Estimator()
id3_model2.fit(label_X_train, label_y_train)

tree_text2 = export.export_text(id3_model2.tree_, feature_names=list(tennis.columns))
print(tree_text2)


outlook <=0.50: 1 (3) 
outlook >0.50
|   humidity <=0.50: 0 (3) 
|   humidity >0.50
|   |   windy <=0.50: 1 (3) 
|   |   windy >0.50
|   |   |   temp <=1.00: 0 (1) 
|   |   |   temp >1.00: 1 (1) 



Evaluate the model to training data with Accuracy and F1 Metrics

In [39]:
predict_train2 = id3_model2.predict(label_X_train)

print("Accuracy: ", accuracy_score(predict_train2,label_y_train))
print("F1 score: ", f1_score(predict_train2,label_y_train))

Accuracy:  1.0
F1 score:  1.0


Evaluate the model to test data with using Accuracy and F1 Metrics

In [40]:
predict_test2 = id3_model2.predict(label_X_valid)

print("Accuracy: ", accuracy_score(predict_test2,label_y_valid))
print("F1 score: ", f1_score(predict_test2,label_y_valid))

Accuracy:  0.6666666666666666
F1 score:  0.6666666666666666


### KMeans

Define KMeans and then train the dataset.

In [41]:
from sklearn.cluster import KMeans

# Determine how many clusters
n_clst = 2
# Define the KMeans
kmeans_model_2 = KMeans(n_clusters = n_clst, random_state=1)
kmeans_model_2.fit(label_X_train)

KMeans(n_clusters=2, random_state=1)

Predict and produce clusters.

In [42]:
predict_train = kmeans_model_2.predict(label_X_train)
predict_test = kmeans_model_2.predict(label_X_valid)

Evaluate the training data now

In [43]:
# Find out each cluster belong with which class

# First clusterization using first prediction (k_means_train)
# Calculate the accuracy and F1
acc_train = accuracy_score(label_y_train, predict_train)
f1_train = f1_score(label_y_train, predict_train)
print(f"First clusterization accuracy: {acc_train}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_train = pd.Series(predict_train)
fix = switched_train.apply(switch)
acc_train1= accuracy_score(label_y_train, fix)
f1_train2= f1_score(label_y_train, fix)
print(f"Second clusterization accuracy: {acc_train1}")
print(f"Second clusterization F1: {f1_train2}")

First clusterization accuracy: 0.36363636363636365
First clusterization F1: 0.4615384615384615
Second clusterization accuracy: 0.6363636363636364
Second clusterization F1: 0.6666666666666666


First clusterization is better, so we take the metrics as the final metrics. Evaluate the validation data now

In [44]:
# First clusterization
acc_valid = accuracy_score(label_y_valid, predict_test)
f1_valid = f1_score(label_y_valid, predict_test)
print(f"First clusterization accuracy: {acc_valid}")
print(f"First clusterization F1: {f1_train}")

# Second clusterization: switch 0 to 1
switched_valid = pd.Series(predict_test).apply(switch)
acc_valid_switch = accuracy_score(label_y_valid, switched_valid)
f1_valid_switch = f1_score(label_y_valid, switched_valid)
print(f"Second clusterization accuracy: {acc_valid_switch}")
print(f"Second clusterization F1: {f1_valid_switch}")

First clusterization accuracy: 0.0
First clusterization F1: 0.4615384615384615
Second clusterization accuracy: 1.0
Second clusterization F1: 1.0


Same as when we evaluate with training data, the first clusterization is better and then will be our final metrics score.

### LogisticRegression

Scaling the dataset.

In [39]:
# # Scaling if needed
# from sklearn.preprocessing import StandardScaler, MinMaxScaler
# scaler = MinMaxScaler()

# label_X_train = scaler.fit_transform(label_X_train)
# label_X_valid = scaler.fit_transform(label_X_valid)

Modelling

In [45]:
model = LogisticRegression()

model.fit(label_X_train,label_y_train)

LogisticRegression()

Predict the train data and evaluate the prediction with Accuracy and F1 score.

In [46]:
predict_train = model.predict(label_X_train)

acc = accuracy_score(label_y_train, predict_train)
f1 = f1_score(label_y_train, predict_train)

print("Accuracy: ", acc)
print("F1 score: ", f1)

Accuracy:  0.8181818181818182
F1 score:  0.8750000000000001


tanpa scaling:
- Accuracy:  0.8181818181818182
- F1 score:  0.8750000000000001

StandardScaler:
- Accuracy:  0.9090909090909091
- F1 score:  0.9333333333333333

MinMaxScaler:
- Accuracy:  0.8181818181818182
- F1 score:  0.8750000000000001

Predict the test data and evaluate the prediction with Accuracy and F1 score.

In [47]:
predict_test = model.predict(label_X_valid)

acc = accuracy_score(label_y_valid, predict_test)
f1 = f1_score(label_y_valid, predict_test)

print("Accuracy: ", acc)
print("F1 score: ", f1)

Accuracy:  0.6666666666666666
F1 score:  0.8


Gaada bedanya pake scaling atau engga
- Accuracy:  0.6666666666666666
- F1 score:  0.8

### Neural_network

### SVM

Define the SVM (Support Vector Machines) constructor

In [49]:
svm = SVC(random_state=1)

Train the dataset

In [54]:
svm.fit(label_X_train, y_train_tennis.values.ravel())
# .values will give the values in an array (shape: (n,1)
# .ravel will convert that array shape to (n, )
# y_train_tennis.values.ravel()

SVC(random_state=1)

Predict the data train and the data test

In [64]:
pred_svm_train_tennis = svm.predict(label_X_train)
pred_svm_val_tennis = svm.predict(label_X_valid)
pred_svm_train_tennis
# pred_svm_val_tennis


array(['yes', 'yes', 'yes', 'no', 'yes', 'no', 'yes', 'yes', 'yes', 'yes',
       'yes'], dtype=object)

Evaluate the prediction of data train with using Accuracy and F1 metrics

In [68]:
acc_score_svm = accuracy_score(y_train_tennis, pred_svm_train_tennis)
# must define the positive label = yes
f1_score_svm = f1_score(y_train_tennis, pred_svm_train_tennis, pos_label="yes")
print("Accuracy : ",acc_score_svm)
print("F1 : ",f1_score_svm)

Accuracy :  0.8181818181818182
F1 :  0.8750000000000001


Evaluate the prediction of data validation with using Accuracy and F1 metrics

In [70]:
acc_score_svm = accuracy_score(y_valid_tennis, pred_svm_val_tennis)
# must define the positive label = yes
f1_score_svm = f1_score(y_valid_tennis, pred_svm_val_tennis, pos_label="yes")
print("Accuracy : ",acc_score_svm)
print("F1 : ",f1_score_svm)

Accuracy :  1.0
F1 :  1.0


# Analysis accuracy and F1 score for each model